Lake/Load/Materialize.lean

```diff
@@ -26,89 +27,75 @@ def cloneGitPkg (repo : GitRepo) (url : String) (rev? : Option String) : LogIO P
+        logInfo s!"{name}: URL has changed; you might need to delete {repo.dir} manually"
+        updateGitPkg repo rev?
       else
-        cloneGitPkg repo entry.url entry.rev
-      return entry
+        logInfo s!"{name}: URL has changed; deleting {repo.dir} and cloning again"
+        IO.FS.removeDirAll repo.dir
+        cloneGitPkg repo url rev?
   else
-    if (← repo.dirExists) then
-      logInfo s!"{name}: no manifest entry; deleting {dir} and cloning again"
-      IO.FS.removeDirAll dir
-      cloneGitPkg repo url rev?
-    else
-      cloneGitPkg repo url rev?
+    cloneGitPkg repo url rev?
+
+def updateSource (relParentDir packagesDir : FilePath) (name : String) (source : Source) : LogIO PackageEntry :=
+  match source with
+  | .path dir => return .path name (relParentDir / dir)
+  | .git url inputRev? subDir? => do
+    let dir := packagesDir / name
+    let repo := GitRepo.mk dir
+    updateGitRepo repo url inputRev? name
     let rev ← repo.headRevision
-    return {name, url, rev, inputRev? := rev?}
+    return .git name url rev inputRev? subDir?
 
 structure MaterializeResult where
   pkgDir : FilePath
+  relPkgDir : FilePath
   remoteUrl? : Option String
   gitTag? : Option String
-  manifestEntry? : Option PackageEntry
+  manifestEntry : PackageEntry
+  deriving Repr, Inhabited
 
 /--
-Materializes a `Dependency`, downloading nd/or updating it as necessary.
-Local dependencies are materialized relative to `localRoot` and remote
-dependencies are stored in `packagesDir`.
+Materializes a package entry, cloning and/or checkout it out as necessary.
 -/
-def materializeDep (packagesDir localRoot : FilePath)
-(dep : Dependency) (manifestEntry? : Option PackageEntry)
-: LogIO MaterializeResult :=
-  match dep.src with
-  | Source.path dir =>
-    return ⟨localRoot / dir, none, none, none⟩
-  | Source.git url rev? subDir? => do
-    let name := dep.name.toString (escape := false)
-    let gitDir := packagesDir / name
-    let entry? ← materializeGitPkg name gitDir url rev? manifestEntry?
-    let pkgDir :=
-      match subDir? with
-      | some subDir => gitDir / subDir
-      | none => gitDir
-    let tag? ← GitRepo.mk gitDir |>.findTag?
-    let url? := Git.filterUrl? url
-    return ⟨pkgDir, url?, tag?, entry?⟩
+def materializePackageEntry (rootPkg : Package) (manifestEntry : PackageEntry) : LogIO MaterializeResult :=
+  match manifestEntry with
+  | .path _name pkgDir =>
+    return {
+      pkgDir := rootPkg.dir / pkgDir
+      relPkgDir := pkgDir
+      remoteUrl? := none
+      gitTag? := none
+      manifestEntry
+    }
+  | .git name url rev _inputRev? subDir? => do
+    let relGitDir := rootPkg.config.packagesDir / name
+    let gitDir := rootPkg.dir / relGitDir
+    let repo := GitRepo.mk gitDir
+    /-
+    Do not update (fetch remote) if already on revision
+    Avoids errors when offline e.g. [leanprover/lake#104][104]
+    [104]: https://github.com/leanprover/lake/issues/104
```

Comment:
This is my fault from the original source, but there should be a blank line between the the reference and link (plus periods).

```suggestion
    Do not update (fetch remote) if already on revision.
    Avoids errors when offline e.g. [leanprover/lake#104][104].
    
    [104]: https://github.com/leanprover/lake/issues/104
```